src/main/scala/com/fulcrumgenomics/umi/ConsensusCaller.scala

```diff
@@ -87,6 +87,89 @@ case class ConsensusRead(bases: Array[Byte], quals: Array[Byte]) {
   def baseString = new String(bases)
 }
 
+/** Contains methods to help perform perform various consensus calling operations efficiently. Requires some upfront
+  * computation of various probabilities. */
+trait ConsensusCallerMath {
+
+  /** The Phred-scaled error rate for an error post the UMIs have been integrated. */
+  protected def errorRatePostUmi: PhredScore
+
+  // Some useful constants
+  protected val LogThree      = LogProbability.toLogProbability(3.0)
+  protected val LogFourThirds = LogProbability.toLogProbability(4.0) - LogThree
+
+  /** Computes the probability of seeing an error in the base sequence if there are two independent error processes.
+    * We sum three terms:
+    * 1. the probability of an error in trial one and no error in trial two: Pr(A=Error, B=NoError).
+    * 2. the probability of no error in trial one and an error in trial two: Pr(A=NoError, B=Error).
+    * 3. the probability of an error in both trials, but when the second trial does not reverse the error in first one, which
+    *    for DNA (4 bases) would only occur 2/3 times: Pr(A=x->y, B=y->z) * Pr(x!=z | x!=y, y!=z, x,y,z \in {A,C,G,T})
+    */
+  def probabilityOfErrorTwoTrials(prErrorTrialOne: LogProbability, prErrorTrialTwo: LogProbability): LogProbability = {
+    if (prErrorTrialOne < prErrorTrialTwo) probabilityOfErrorTwoTrials(prErrorTrialTwo, prErrorTrialOne)
+    else if (prErrorTrialOne - prErrorTrialTwo >= 6) prErrorTrialOne // a simple approximation since prErrorTrialOne will dominate
+    else {
+      // f(X, Y) = X(1-Y) + (1-X)Y + 2/3*XY
+      //         = X - XY + Y - XY + 2/3*XY
+      //         = X + Y - 2XY + 2/3*XY
+      //         = X + Y + XY*(2/3 - 6/3)
+      //         = X + Y - 4/3*XY
+      val term1 = LogProbability.or(prErrorTrialOne, prErrorTrialTwo) // X + Y
+      val term2 = LogFourThirds + prErrorTrialOne + prErrorTrialTwo // 4/3*XY
+      LogProbability.aOrNotB(term1, term2)
+    }
+  }
+
+  /** Pre-computes the the log-scale probabilities of an error for each a phred-scaled base quality from 0-100. */
+  protected lazy val phredToLogProbError: Array[Double] = Range(0, 100).toArray.map(p => {
```

Comment:
Turning these into lazy vals is going to add a lot of overhead - previously use of these was a direct access into an array, now each access into the array will be via a method call that does an initialization check (with, I think, synchronization overhead).